src/applications/model/gaming-streaming-server.h

```diff
@@ -106,32 +107,36 @@ class GamingStreamingServer : public Application
    * \param packetSize A random number generator for packet size
    * \param interArrivalTime A random number generator for inter-arrival time
    */
-  void AddNewTraffic (Ptr<RandomVariableStream> packetSize, Ptr<RandomVariableStream> interArrivalTime);
+  void AddNewTrafficStream (Ptr<RandomVariableStream> packetSize, Ptr<RandomVariableStream> interArrivalTime);
 
   virtual void StartApplication (void);
   virtual void StopApplication (void);
 
 protected:
-  /**
-   * \brief Generate and send packets based on the different random
-   * distribution for packet sizes and inter-arrival times
-   *
-   * \param idx index of stream within the streams vector
-   */
-  void Send (int idx);
+  virtual void DoDispose (void)  override;
 
   /**
    *  Initialize the parameters of different streams
    */
   virtual void InitializeStreams () = 0;
 
 private:
+  struct TrafficStream;
+
+  /**
+   * \brief Generate and send packets based on the   random
+   * distribution for packet sizes and inter-arrival times
+   *
+   * \param traffic the pointer to traffic stream defined by TrafficStream struct
+   */
+  void Send (TrafficStream* traffic);
```

Comment:
I schedule the `Send` method in the `Simulator:: Schedule`, and the problem is `Schedule` doesn't accept _passed by reference arguments_ ([Here](https://groups.google.com/forum/#!topic/ns-3-users/OuJREj-Qt0k)), and make a local copy of `traffic`.
So at the end of simulation I don't have access to cancel the `sendEvents`. (The problem I told you once.)

Reply:
Technically both of `TrafficStream*` and `TrafficStream&` are called by reference. but I don't know why to buy `&`, I'm not getting the proper results.

Reply:
the link refers to ns-3.12, we are dealing with ns-3.26 (i think), is that still the case?
If so, use `std::vector<Ptr<TrafficStream> > m_trafficStreams;`, instead, to avoid raw pointers.

Reply:
Yes, it's still a case in the new version.
Is it possible to use smart pointer (Ptr<>) on a struct? if yes, how?
I tried and failed. It seems only classes inherited from `SimpleRefCount` can be passed as an argument to Ptr<>.

Reply:
I solved the problem by driving the `TrafficStream` struct from the object class. I'm not sure how much it is stupid, but it works.